modules/01_Theory.md

```diff
@@ -0,0 +1,502 @@
+
+   | TypeScript file extension | Declaration file extension | JavaScript file extension |
+   | ------------------------- | -------------------------- | ------------------------- |
+   | `.ts`                     | `.d.ts`                    | `.js`                     |
+   | `.tsx`                    | `.d.ts`                    | `.js`                     |
+   | `.mts`                    | `.d.mts`                   | `.mjs`                    |
+   | `.cts`                    | `.d.cts`                   | `.cjs`                    |
+   |                           | `.d.*.ts`                  | `.*`                      |
+   
+   The last row expresses that non-JS files can be typed with the `allowArbitraryExtensions` compiler option to support cases where the module system supports importing non-JS files as JavaScript objects. For example, a file named `styles.css` can be represented by a declaration file named `styles.d.css.ts`.
+
+3. **Every module specifier that resolves to a declaration file in TypeScript must resolve to its JavaScript file counterpart in the runtime.** The easiest way to conform to this rule is to make every declaration file a direct sibling of its corresponding JavaScript file, with the same name. (Or, in the case of a package on DefinitelyTyped, the files in the `@types` package should be a 1:1 mirror of the files in the implementation package, even though they can’t be direct siblings on disk.) This is the default behavior of `tsc --declaration`, so it’s easiest to understand this rule by looking at a violation of it:
+   ```json5
+   // @Filename: /node_modules/greetings/package.json
+   {
+    "name": "greetings",
+    "main": "greetings.js", // Assume these files exist.
+    "types": "index.d.ts"   // The declaration file name is different!
+   }
+   ```
+   ```ts
+   // @moduleResolution: node16
+   // @Filename: /main.ts
+   import { sayHello } from "greetings";         // ✅ TS: index.d.ts, Node.js: greetings.js.
+   import { sayGoodbye } from "greetings/index"; // ❌ TS: index.d.ts, Node.js: 💥
+   ```
+   It might have seemed fine for the declaration file to be named `index.d.ts` while the JavaScript file was named `greetings.js` since these filenames are hidden behind `"types"` and `"main"` for root package-name imports. But according to Node.js’s module resolution rules, we’re also allowed to access individual files in this package by name, where `"types"` and `"main"` don’t apply. There are ways to put declaration files in dedicated directories in published packages, but it requires extra care (and configuration) to ensure that the correct mapping from declaration file location to JavaScript file location is always used. It’s highly recommended to colocate declaration files with their implementation files until there’s a good reason not to.
+
+### A tour of common module resolution features
```

Comment:
Much or all of this will get deleted or replaced with brief mentions and links to the reference pages